File: VB/HowToCustomizeNewActionItemsList.Module.Win/MyController.vb
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Win.SystemModule
Imports DevExpress.ExpressApp.SystemModule


Partial Public Class MyController
   Inherits WinNewObjectViewController
   'Subscribe the required events
   Protected Overrides Sub OnActivated()
      'Get the ShowNavigationItemController,
      'then get its ShowNavigationItemAction and subscribe the SelectedItemChanged event
      AddHandler Frame.GetController(Of ShowNavigationItemController)().ShowNavigationItemAction.SelectedItemChanged, AddressOf ShowNavigationItemAction_SelectedItemChanged
      AddHandler CollectCreatableItemTypes, AddressOf MyController_CollectCreatableItemTypes
      AddHandler CollectDescendantTypes, AddressOf MyController_CollectDescendantTypes
      MyBase.OnActivated()
   End Sub
   Private Sub MyController_CollectDescendantTypes(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs)
      CustomizeList(e.Types)
   End Sub
   Private Sub MyController_CollectCreatableItemTypes(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs)
      CustomizeList(e.Types)
   End Sub
   Private Sub ShowNavigationItemAction_SelectedItemChanged(ByVal sender As Object, ByVal e As EventArgs)
      Me.UpdateActionState()
   End Sub
   Public Sub CustomizeList(ByVal types As ICollection(Of Type))
      'Get the ShowNavigationItemController, then get its ShowNavigationItemAction
      Dim showNavigationItemAction As SingleChoiceAction = Frame.GetController(Of ShowNavigationItemController)().ShowNavigationItemAction
      'Get the item selected in the navigation control
      Dim selectedItem As ChoiceActionItem = showNavigationItemAction.SelectedItem
      Dim currentGroup As ChoiceActionItem = Nothing
      If Not selectedItem Is Nothing Then
         'Get the selected item's parent group
         currentGroup = selectedItem.ParentItem
         Dim unusableTypes As List(Of Type) = New List(Of Type)()
         'Collect the types that must be deleted
         For Each type As Type In types
            Dim deletionRequired As Boolean = True
            For Each item As ChoiceActionItem In currentGroup.Items
               Dim shortcut As ViewShortcut = TryCast(item.Data, ViewShortcut)
               If shortcut.ViewId = Application.FindListViewId(type) Then
                  deletionRequired = False
               End If
            Next item
            If deletionRequired = True Then
               unusableTypes.Add(type)
            End If
         Next type
         'Remove the collected types
         For Each type As Type In unusableTypes
            types.Remove(type)
         Next type
      End If
   End Sub
   'Unsubscribe from the events
    Protected Overrides Sub OnDeactivated()
        RemoveHandler Frame.GetController(Of ShowNavigationItemController)().ShowNavigationItemAction.SelectedItemChanged, AddressOf ShowNavigationItemAction_SelectedItemChanged
        RemoveHandler CollectCreatableItemTypes, AddressOf MyController_CollectCreatableItemTypes
        RemoveHandler CollectDescendantTypes, AddressOf MyController_CollectDescendantTypes
        MyBase.OnDeactivated()
    End Sub
End Class

File: CS/HowToCustomizeNewActionItemsList.Module.Win/MyController.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;

using DevExpress.ExpressApp.Win.SystemModule;
using DevExpress.ExpressApp.SystemModule;

namespace HowToCustomizeNewActionItemsList.Module.Win {
public partial class MyController : WinNewObjectViewController {
	//Subscribe the required events
	protected override void OnActivated() {
		//Get the ShowNavigationItemController,
		//then get its ShowNavigationItemAction and subscribe the SelectedItemChanged event
		Frame.GetController<ShowNavigationItemController>().ShowNavigationItemAction.SelectedItemChanged +=
			new EventHandler(ShowNavigationItemAction_SelectedItemChanged);
		CollectCreatableItemTypes += new EventHandler<DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs>(MyController_CollectCreatableItemTypes);
		CollectDescendantTypes += new EventHandler<DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs>(MyController_CollectDescendantTypes);
		base.OnActivated();
	}
	void MyController_CollectDescendantTypes(object sender, DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs e) {
		CustomizeList(e.Types);
	}
	void MyController_CollectCreatableItemTypes(object sender, DevExpress.ExpressApp.SystemModule.CollectTypesEventArgs e) {
		CustomizeList(e.Types);
	}
	void ShowNavigationItemAction_SelectedItemChanged(object sender, EventArgs e) {
		this.UpdateActionState();
	}
	public void CustomizeList(ICollection<Type> types) {
		//Get the ShowNavigationItemController, then get its ShowNavigationItemAction
		SingleChoiceAction showNavigationItemAction =
			Frame.GetController<ShowNavigationItemController>().ShowNavigationItemAction;
		//Get the item selected in the navigation control
		ChoiceActionItem selectedItem = showNavigationItemAction.SelectedItem;
		ChoiceActionItem currentGroup = null;
		if(selectedItem != null) {
			//Get the selected item's parent group
			currentGroup = selectedItem.ParentItem;
			List<Type> unusableTypes = new List<Type>();
			//Collect the types that must be deleted
			foreach(Type type in types) {
				bool deletionRequired = true;
				foreach(ChoiceActionItem item in currentGroup.Items) {
					ViewShortcut shortcut = item.Data as ViewShortcut;
					if(shortcut.ViewId == Application.FindListViewId(type)) {
						deletionRequired = false;
					}
				}
				if(deletionRequired == true)
					unusableTypes.Add(type);
			}
			//Remove the collected types
			foreach(Type type in unusableTypes)
				types.Remove(type);
		}
	}
	//Unsubscribe from the events
	protected override void OnDeactivated() {
		Frame.GetController<ShowNavigationItemController>().ShowNavigationItemAction.SelectedItemChanged -=
			new EventHandler(ShowNavigationItemAction_SelectedItemChanged);
		CollectCreatableItemTypes -= new EventHandler<CollectTypesEventArgs>(MyController_CollectCreatableItemTypes);
		CollectDescendantTypes -= new EventHandler<CollectTypesEventArgs>(MyController_CollectDescendantTypes);
        base.OnDeactivated();
	}
}
}
